### ISSUE TITLE
Span improvements

### ISSUE THREAD

**Issue by kaleidawave** [kaleidawave/ezno#17]
All AST contains information about which slice and the position in the slice it was parsed from through [Span](https://docs.rs/source-map/latest/source_map/struct.Span.html). Currently:
- Spans are start byte index in the string, end byte index in the string.
- Binary expressions (and some others combined structures) don't have position information, instead they combine the left and right spans.

There are opportunities for improvements here, which may require changes to [`source-map`](https://docs.rs/source-map).

#### Changes with lexing:
- Token sends a Span. There are two redundant pieces of information
  - SourceId. This is sent the same for every token. It should be on ParsingState instead
  - Is an end position needed. This maybe could be worked out from a start + length of token (token length won't change so isn't an issue). 
  - Instead tokens would be of type `Token<TSXToken, /* start */ usize>`. And to construct a AST position it would be `Span { start: token_start, end: token_start + token_length, source_id: state.current_source_id }`

#### Changes to the AST, parsing and spans:

- Are byte slice based spans a good idea. Do line column spans offer benefits?
   - Some of the #1 could benefit from knowing about line breaks, while possible with slice based one it's a bit overhead
- Should the end marker be relative or absolute?
- Do computed positions cause issues? Are errors ever produce on modified AST? Aka swapping binary operands leads to a structure with broken position information.
- For literals, do they need a end byte marker. Could be more efficient if the end is computed via start + literal length? Does that suffer from the problems from directly above.

#### Libraries that use bytewise start-end spans:
- [codespan-reporting](https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/diagnostic/struct.Label.html#structfield.range)

#### Libraries that use line-column spans:
- Source maps are line-column based
- LSP and so [LSP types](https://docs.rs/lsp-types/0.94.0/lsp_types/struct.Range.html)
- Rollup reporting (used by the JS WASM edition) is currently line-column

**Comment by kaleidawave** [comment]
Thanks for the comment, some good discussion/brainstorm 💡.

> The algorithm is pretty simple actually:
> 
> * mark the token if it's on a new line:

Ah that is another interesting way "Add a marker to the token data **after a new line is encountered**". Will add that to the list!

My next question would be: if the token position data from lexing was line-column based then instead of a `on_new_line` boolean, could new lines be found where necessary by something like: `previous_ast.end_line() != current_token.start_line()` (which would very minorly reduce the size of a token data)??

---

> > Do computed positions cause issues? Are errors ever produce on modified AST? Aka swapping binary operands leads to a structure with broken position information
> 
> What are you trying to achieve here?

Some time ago I made certain AST like binary expressions lazily compute there span using by combining their left and right (which is why it returns a `Cow<Span>`):
https://github.com/kaleidawave/ezno/blob/f64028b23b07ce17432f43438d8e3a58be990397/parser/src/expressions/mod.rs#L253-L257

I think I was trying to reduce the AST size but I now realize that this is not great 🤣. If I swap the operands it will then cause a inverted span. Found this out here, which is why the diagnostic positioning needs to be before the `std::mem::swap` call:
https://github.com/kaleidawave/ezno/blob/f64028b23b07ce17432f43438d8e3a58be990397/src/temp.rs#L103-L116 

I probably need to backtrack on that decision. While it's duplicate information on an unmodified AST, it causes chaos after modifying the AST. And as I have now learnt the size of AST is the maximum variant, so it probably has no effect. I should properly check data sizes before making these changes.

---

> For literals, do they need a end byte marker.

If you have a token such as `Token(TSXToken::IdentLiteral("console".into()), TokenData { start: 100, end: 107 })`. Does it need `end`, as it is `start + "console".len()`? The lexer doesn't process escape sequences and leaves them in their source form. All other kinds of tokens have known source size, right? Maybe going a bit extreme here but wondering the absolute smallest Token necessary 😂
